README.md
# sovereign: State-Dependent Empirical Analysis  

<!-- badges: start -->
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
[![CRAN status](https://www.r-pkg.org/badges/version/sovereign)](https://CRAN.R-project.org/package=sovereign)
[![Lifecycle: stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![codecov](https://codecov.io/gh/tylerJPike/sovereign/branch/main/graph/badge.svg?token=WXLWR6H93B)](https://codecov.io/gh/tylerJPike/sovereign)
[![R-CMD-check](https://github.com/tylerJPike/sovereign/workflows/R-CMD-check/badge.svg)](https://github.com/tylerJPike/sovereign/actions)
<!-- badges: end -->

The sovereign package introduces a set of tools for state-dependent empirical analysis through both VAR- and local projection-based state-dependent forecasts, impulse response functions, forecast error variance decompositions, and historical decompositions.    

See the sovereign package [website](https://tylerjpike.github.io/sovereign/) for examples and documentation. 

----

## Installation 

One may install `sovereign` through the package’s [GitHub](https://github.com/tylerJPike/sovereign) page with

    devtools::install_github('tylerJPike/sovereign')

or through [CRAN](https://CRAN.R-project.org/package=sovereign) via

    install.packages("sovereign")

----

## Available Tools  

Unsupervised Regime Assignment
1. random forest  
2. k-means clustering  
3. EM 
4. Bai & Perron (2003)  

Local Projections (LP)
1. direct projection forecasting  
1. impulse responses

Vector Auto-Regression (VAR), Structural VAR (SVAR),   
and external instrument SVAR (Proxy-SVAR)  
1. recursive forecasting
2. impulse responses
3. forecast error variance decomposition
4. historical error decomposition


----

## Basic Workflow 
    # load packages
    library(sovereign)         # analysis
    library(dplyr)             # general cleaning
    library(lubridate)         # date functions

    #-------------------------------------------
    # create data
    #-------------------------------------------
    # pull and prepare data from FRED
    quantmod::getSymbols.FRED(
        c('UNRATE','INDPRO','GS10'), 
        env = globalenv())

    Data = cbind(UNRATE, INDPRO, GS10)

    Data = data.frame(Data, date = zoo::index(Data)) %>%
        filter(lubridate::year(date) >= 1990,
               lubridate::year(date) <  2020) %>% 
        na.omit()

    # create a regime explicitly   
    Data.threshold = Data %>%
        mutate(mp = if_else(GS10 > median(GS10), 1, 0))

    #------------------------------------------
    # learn regimes
    #------------------------------------------
    # assign regimes based on unsurpervised kmeans
    #  (will not be used further)
    regimes = 
        regimes(
            data = Data, 
            regime.n = 3, 
            engine = 'kmeans')

    #------------------------------------------
    # single-regime var
    #------------------------------------------
    # estimate VAR
    # (using IC lag selection and short-term 
    #  impact restrictions, i.e. calculate structurals
       errors via Cholesky decomposition)
    var =
        VAR(
            data = Data,
            horizon = 10,
            freq = 'month',
            lag.ic = 'BIC',
            lag.max = 4)

    # if impacts of the COVID shock are present,
    #  in the user's data, then the user can 
    #  implement the Lenza-Primiceri COVID volatility
    #  correction for VARs: 
    # var = covid_volatility_correction(var)

    # plot forecasts
    plot_forecast(var$forecasts[[1]])

    # plot residuals
    plot_error(var$residuals[[1]])

    # estimate IRF
    irf =
        IRF(
            var,
            bootstraps.num = 10,
            CI = c(0.05,0.95))

    # plot IRF
    plot_irf(irf)

    # estimate forecast error variance decomposition
    fevd =
        FEVD(
            var,
            horizon = 10)

    # plot FEVD
    plot_fevd(fevd)

    # estimate historical decomposition
    hd = HD(var)
    
    # plot HD
    plot_hd(hd)

    #-------------------------------------------
    # multi-regime var
    #-------------------------------------------
    # estimate multi-regime VAR
    var =
        RVAR(
            data = Data.threshold,
            regime = 'mp',
            p = 1,
            horizon = 1,
            freq = 'month')
    
    # estimate IRF
    rvar.irf =
        IRF(
            rvar,
            horizon = 10,
            bootstraps.num = 10,
            CI = c(0.05,0.95))

    # plot IRF
    # regime 1: low interest rates
    plot_irf(rvar.irf[[1]])
    # regime 2: high interest rates
    plot_irf(rvar.irf[[2]])

    # estimate forecast error variance decomposition
    rvar.fevd =
        FEVD(
            rvar,
            horizon = 10)

    # plot FEVD
    # regime 1: low interest rates
    plot_fevd(rvar.fevd[[1]])
    # regime 2: high interest rates
    plot_fevd(rvar.fevd[[2]])

    # estimate HD
    rvar.hd = HD(rvar)

    #-------------------------------------------
    # single-regime local projections
    #-------------------------------------------
    # estimate single-regime forecasts 
    #  (one or multiple horizons may be estimated)
    lp = 
        LP(
            data = Data,
            p = 1,
            horizon = 1,
            freq = 'month')

    # estimate single-regime IRF
    lp.irf = IRF(lp)

    # plot IRF
    plot_irf(lp.irf)

    #-------------------------------------------
    # multi-regime local projections
    #-------------------------------------------
    # estimate multi-regime IRF
    rlp = 
        RLP(
            data = Data,
            regime = 'mp',
            p = 1,
            horizon = 1,
            freq = 'month')

    # estimate multi-regime IRF
    rlp.irf = IRF(rlp)

    # plot IRF
    # regime 1: low interest rates
    plot_irf(rlp.irf[[1]])
    # regime 2: high interest rates
    plot_irf(rlp.irf[[2]])
    

---
## Contact
If you should have questions, concerns, or wish to collaborate, please contact [Tyler J. Pike](https://tylerjpike.github.io/)


NEWS.md
# Sovereign: State-Dependent Empirical Analysis  

## Version 1.2.0

- Major updates
    - VARs may now be estimated as Proxy-SVARs
    - IRF functions now support
      - Choice of bootstrapping techniques 
        - standard residual resampling
        - Wild bootstrapping
      - Choice of structural shocks 
        - no structural restrictions, 
        - short-term restrictions (via Cholesky decomposition), 
        - instrumental variable estimation  
        - combination of instrument variable and short-term restrictions  
      - Parallel processing for greater computational efficiency in computing bootstrapped confidence intervals   
      
- Minor updates
  - IRF plots use darker blue for CI
  - Analysis function wrappers implemented for IRFs, FEVDs, and HDs
  - Updated model documentation with academic references and corrected minor errors 
   
- Bug fixes
    - plot_irf() plots all targets by default 
    - test-threvhold_var.R changed to test-regime_var.R
    - Fixed IRF results format listed in documentation

## Version 1.1.0   
(2021-06-01)

- Major updates
    - Include COVID shock correction a la Lenza and Primiceri (2020)
    - Include historical decomposition of VAR shocks
      
- Minor updates
    - Create object classes for LP and VAR output  
    - n.lag no longer requires a 'date' column 
    - Specify structural assumptions in IRFs  
    - Clarify model specification of RVAR  

- Bug fixes
    - Convert errorCondition() to stop()   

## Version 1.0.0 
(2021-04-02)

- Initial release 


R/regimes.R
#------------------------------------------
# Function to assign regimes via
#   unsupervised machine learning
#------------------------------------------
#' Identify regimes via unsupervised ML algorithms
#'
#' Regime assignment (clustering) methods available include the
#' [unsupervised random forest](https://www.rdocumentation.org/packages/randomForest/versions/4.6-14/topics/randomForest),
#' [k-mean clustering](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/kmeans),
#' Fraley and Raftery Model-based clustering [EM algorithm](https://www.rdocumentation.org/packages/mclust/versions/5.4.7/topics/Mclust),
#' and the [Bai & Perron (2003)](https://www.rdocumentation.org/packages/strucchange/versions/1.5-2/topics/breakpoints) method for simultaneous estimation of multiple breakpoints.
#'
#' @param data                  data.frame, matrix, ts, xts, zoo: Endogenous regressors
#' @param method                string: regime assignment technique ('rf', 'kmeans', 'EM', or 'BP)
#' @param regime.n              int: number of regimes to estimate (applies to kmeans and EM)
#'
#' @return `data` as a data.frame with a regime column assigning rows to mutually exclusive regimes
#'
#' @examples
#' \donttest{
#'
#'  # simple time series
#'  AA = c(1:100) + rnorm(100)
#'  BB = c(1:100) + rnorm(100)
#'  CC = AA + BB + rnorm(100)
#'  date = seq.Date(from = as.Date('2000-01-01'), by = 'month', length.out = 100)
#'  Data = data.frame(date = date, AA, BB, CC)
#'
#'  # estimate reigme
#'  regime =
#'   sovereign::regimes(
#'      data = Data,
#'      method = 'kmeans',
#'      regime.n = 3)
#' }
#'
#' @export
regimes = function(
  data,                        # data.frame, matrix, ts, xts, zoo: Endogenous regressors
  method = 'rf',               # string: regime assignment technique ('rf', 'kmeans', 'EM', 'BP')
  regime.n = NULL              # int: number of regimes to estimate (applies to kmeans and EM)
){

  # function warnings
  if(!is.matrix(data) & !is.data.frame(data)){
    stop('data must be a matrix or data.frame')
  }
  if(!is.null(regime.n) & !is.numeric(regime.n)){
    stop('regime.n must be a positive integer or NULL')
  }

  # cast as data frame if ts, xts, or zoo object
  if(stats::is.ts(data) | xts::is.xts(data) | zoo::is.zoo(data)){
    data = data.frame(date = zoo::index(date), data)
  }

  # check for BP method
  if(method == 'bp' & ncol(dplyr::select(data, -date)) > 1){
    stop("The 'BP' method can only use univariate time series to determine regimes.")
  }

  # clean data
  X = dplyr::select(data, -date)
  X.date = data$date

  # assign regimes
  if(method == 'rf'){

    model = randomForest::randomForest(X)
    regime = data.frame(model$votes)
    colnames(regime) = paste0('Prob_regime_',c(1:ncol(regime)))
    regime = apply(X = regime, MARGIN = 1, which.max)
    regime = data.frame(date = X.date, X, regime)

  }else if(method == 'kmeans'){

    if(is.null(regime.n)){regime.n = 2}
    model = stats::kmeans(X, centers = regime.n)
    regime = data.frame(date = X.date, X, regime = model$cluster)

  }else if(method == 'EM'){

    mclustBIC = mclust::mclustBIC
    model = mclust::Mclust(X, G = regime.n)
    regime = data.frame(date = X.date, X, regime = model$classification)

  }else if(method == 'BP'){

    colnames(X) = 'x'
    breakpoints = strucchange::breakpoints(x ~ lag(x), data = X)
    breakpoints$breakpoints

    if(is.na(breakpoints$breakpoints)){
      regimes = data.frame(date = X.date, X, regime = 0)
    }else{
      regimes.val = c(0:length(breakpoints$breakpoints))
      regimes.dates = data$date[c(1, breakpoints$breakpoints)]
      regimes = dplyr::full_join(dplyr::select(data, date), data.frame(date = regimes.dates, regime = regimes.val), by = 'date')
      regimes = tidyr::fill(regimes, regime)
    }

    regime = data.frame(date = X.date, X, regime = regimes$regime)

  }

  # return regime assignments
  return(regime)

}


R/sovle_b.R

#------------------------------------------
# Function to estimate the structural
#  error impact matrix
#
# See Cesa-Bianchi's VAR-Toolbox for
# MATLAB implementation
#
# IV-short implementation based on matlab
# code by Nick von Turkovich and Paco
# Vazquez-Grande
#------------------------------------------
solve_B = function(var, report_iv = TRUE){

  # function variables
  forecast.date = NA

  # set residuals
  if(class(var) == 'VAR'){
    residuals = var$residuals[[1]]
  }else{
    residuals = var$residuals
  }

  if(is.na(var$structure) == TRUE){

    # retrieve reduced-form residuals
    data.residuals = residuals

    # reduced form variance-covariance matrix
    cov.matrix = stats::var(stats::na.omit(dplyr::select(data.residuals, -date, -forecast.date)))

    B = cov.matrix

    return(B)

  }else if(var$structure == 'short'){

    # retrieve reduced-form residuals
    data.residuals = residuals

    # reduced form variance-covariance matrix
    cov.matrix = stats::var(stats::na.omit(dplyr::select(data.residuals, -date, -forecast.date)))

    # take cholesky decomposition
    B = t(chol(cov.matrix))

    return(B)

  }else if(var$structure == 'IV'){

    # retrieve reduced-form residuals
    data.residuals = residuals
    covariates = colnames(dplyr::select(data.residuals, -date, -forecast.date))
    col_to_instrument = var$instrumented

    # retrieve instrument
    data.instrument = var$instrument
    instrument = colnames(dplyr::select(data.instrument, -date))

    # combine data
    data =
      dplyr::inner_join(data.residuals, data.instrument, by = 'date') %>%
      stats::na.omit()

    # first stage least squares
    model.first_stage =
      stats::lm(col_to_instrument ~.,
         data = data %>%
           dplyr::select(col_to_instrument = dplyr::all_of(col_to_instrument), dplyr::all_of(instrument)) %>%
           stats::na.omit())
    p_hat = model.first_stage$fitted.values

    if(report_iv == TRUE){
      print(summary(model.first_stage))
    }

    # second stage least squares
    # to estimate first column of B
    # (automatically scales first entry to 1)
    instrumented_shocks = covariates %>%
      purrr::map_df(.f = function(X){

        model.second_stage = stats::lm(data[,X] ~ p_hat)
        second_stage_beta = model.second_stage$coefficients[2]
        return(second_stage_beta)

      }) %>%
      as.vector()

    # scale size of the shock
    #  see Gertler and Karadi (2015) for background
    #  see Cesa-Bianchi's VAR-Toolbox for MATLAB implementation
    X.demean =
      dplyr::select(data, -date, -forecast.date, -dplyr::all_of(instrument)) %>%
      dplyr::mutate_all(function(X){return(X - mean(X, na.rm = T))}) %>%
      as.matrix()

    sigma_b = 1/(nrow(data) - ncol(var$model$coef) + 1) * t(X.demean) %*% X.demean

    s21s11 = instrumented_shocks[2:length(covariates),] %>% as.matrix()
    S11 = sigma_b[1,1] %>% as.numeric()
    S21 = sigma_b[2:nrow(sigma_b),1] %>%as.vector()
    S22 = sigma_b[2:nrow(sigma_b),2:ncol(sigma_b)]

    Q = (s21s11 * S11)  %*% t(s21s11) - (S21 %*% t(s21s11) + as.matrix(s21s11) %*% t(S21)) + S22
    sp = sqrt( S11 - t(S21 - as.matrix(s21s11) %*% S11) %*% as.matrix(solve(Q) %*%  (S21 - s21s11 * S11)) )

    scaled_instrumented_shocks = instrumented_shocks * as.numeric(sp)

    # prepare B matrix
    B = matrix(0, ncol = (length(covariates) - 1), nrow = length(covariates))
    B = cbind(scaled_instrumented_shocks, B)

    return(B)

  }else if(var$structure == 'IV-short'){

    # align instrument and residuals
    valid_dates =
      dplyr::inner_join(
        dplyr::select(stats::na.omit(residuals), date),
        dplyr::select(stats::na.omit(var$instrument), date),
        by = 'date'
      )

    # set residuals matrix
    residuals = residuals %>%
      dplyr::inner_join(valid_dates, by = 'date') %>%
      dplyr::select(-date, -forecast.date) %>%
      as.matrix()

    residuals = -1*residuals

    # set instrument
    instrument = var$instrument %>%
      data.frame() %>%
      dplyr::inner_join(valid_dates, by = 'date') %>%
      dplyr::select(-date)

    instrument.mean = instrument %>%
      dplyr::mutate_all(mean, na.rm = T)

    # number of observations
    n.obs = nrow(residuals)

    # solve for IV implied impact
    pi = solve(n.obs^(-1) * t(residuals) %*% residuals) %*%
      (n.obs^(-1) * t(residuals) %*% as.matrix(instrument - instrument.mean))

    phi_sq =
      (n.obs^(-1) * t(instrument - instrument.mean) %*% residuals) %*%
      solve( n.obs^(-1) * t(residuals) %*% residuals ) %*%
      ( n.obs^(-1) * t(residuals) %*% as.matrix(instrument - instrument.mean) )

    B1 =
      n.obs^(-1) *
      ( t(residuals) %*% as.matrix(instrument - instrument.mean) ) %*%
      (phi_sq)^(-0.5)

    B = matrix(ncol = ncol(residuals), nrow = ncol(residuals))
    B[,1:ncol(instrument)] = B1
    rownames(B) = colnames(B) = colnames(residuals)

    # solve for orthogonalized structural shock
    model.first_stage = stats::lm(instrument[,1] ~ residuals)
    orthogonal_instrument = instrument - model.first_stage$residuals
    orthogonal_instrument = orthogonal_instrument[,] / stats::sd(orthogonal_instrument[,], na.rm = T)

    shock_matrix = matrix(nrow = nrow(residuals), ncol = ncol(residuals))
    shock_matrix[,1] = orthogonal_instrument

    # reduced form variance-covariance matrix
    sigma = stats::var(residuals)

    # solve additional entries
    # with a lower triangular restriction
    order_sequence = c(1:ncol(residuals))

    for(i in order_sequence){

      Y = residuals[,i]
      X = shock_matrix[,c(1:i)]
      model.second_stage = stats::lm(Y~X)

      if(i != utils::tail(order_sequence,1)){

        B[i,] =
          c( model.second_stage$coef[-1],
             stats::sd(model.second_stage$residuals),
             rep(0, length(order_sequence) - ncol(instrument) - i) )

        shock_matrix[,i+1] = model.second_stage$residuals / stats::sd(model.second_stage$residuals)

      }else{

        B[i,] =  model.second_stage$coef[-1]

      }

    }

    # make diagonal entries positive
    shock_ordering = data.frame(residuals) %>%
      dplyr::select(var$instrumented, dplyr::everything()) %>%
      colnames()
    B.sign = diag(sign(diag(B[shock_ordering,])))
    B = B %*% B.sign

    return(B)

  }else{

    stop('structure must be set to either NA, "short", "IV", or "IV-short".')

  }

}


R/covid_correction.R
#-------------------------------------------------------
# Function to implement COVID volatility correction
#-------------------------------------------------------

# covid correction maximum likelihood
cvc_likelihood = function(theta_hat, Y.pre, Y.post, p, N){

  # set shock scalar
  Y.pre$s = 1

  t = nrow(Y.post)
  j = c(0:(t-1))
  s_decay = rep(1, t) + (theta_hat[3] - 1) * t(theta_hat[4]^(j-2))
  s = c(theta_hat[1:3], s_decay[4:t])

  Y.post$s = s

  Y = dplyr::bind_rows(Y.pre, Y.post)

  # correct the data
  Y.corrected = Y[, !colnames(Y) %in% c('s', 'date')] / Y[,'s']
  Y.corrected = data.frame(Y.corrected)
  Y.corrected$date = Y$date
  Y.corrected$s = Y$s

  # create lags
  Y.corrected = n.lag(Y.corrected, lags = p)
  Y.corrected$s = Y[,'s']
  Y.corrected = stats::na.omit(Y.corrected)

  # set data
  TT = nrow(Y.corrected)
  Y = dplyr::select(Y.corrected, -date, -dplyr::contains('.l')) %>% as.matrix()
  X = dplyr::select(Y.corrected, dplyr::contains('.l')) %>% as.matrix()
  S = Y.corrected$s

  # calculate beta and sigma
  Beta = solve(t(X) %*% X) %*% t(X) %*% Y
  Epsilon = Y - X %*% Beta
  Sigma = (t(Epsilon) %*% Epsilon)/(TT-p)

  # calculate log-likelihood of y |theta
  logL = -( sum(-N*log(s)) - 0.5*(TT-p)* log(det(Sigma)) )

}


#' Lenza-Primiceri Covid Shock Correction
#'
#' Implement the deterministic volatility correction method of Lenza, Michele
#' and Giorgio Primiceri "How to Estimate a VAR after March 2020" (2020) [[NBER Working Paper](https://www.nber.org/papers/w27771)].
#' Correction factors are estimated via maximum likelihood.
#'
#' @param var       VAR object
#' @param theta_initial   double: four element vector with scaling parameters, theta in Lenza and Primiceri (2020)
#'
#' @return var object
#'
#' @seealso [VAR()]
#' @seealso [var_irf()]
#' @seealso [var_fevd()]
#' @seealso [var_hd()]
#'
#' @examples
#' \donttest{
#'
#'  # simple time series
#'  AA = c(1:100) + rnorm(100)
#'  BB = c(1:100) + rnorm(100)
#'  CC = AA + BB + rnorm(100)
#'  date = seq.Date(from = as.Date('2018-01-01'), by = 'month', length.out = 100)
#'  Data = data.frame(date = date, AA, BB, CC)
#'
#'  # estimate VAR
#'  var =
#'    sovereign::VAR(
#'      data = Data,
#'      horizon = 10,
#'      freq = 'month',
#'      lag.ic = 'BIC',
#'      lag.max = 4)
#'
#' # correct VAR for COVID shock
#' var = sovereign::covid_volatility_correction(var)
#'
#' # impulse response functions
#' var.irf = sovereign::var_irf(var)
#'
#' # forecast error variance decomposition
#' var.fevd = sovereign::var_fevd(var)
#'
#' # historical shock decomposition
#' var.hd = sovereign::var_hd(var)
#'
#' }
#'
#' @export

covid_volatility_correction = function(
  var,                                        # VAR object
  theta_initial = c(5, 2, 1.5, 0.8)           # double: four element vector with scaling parameters, theta in Lenza and Primiceri (2020)
){

  # prepare samples -----------------------------

  # set model variables
  p = var$model$p
  N = nrow(var$model$coef)
  freq = var$model$freq
  Y = var$data

  # pre-COVID shock
  if(freq == 'month'){

    Y.pre = dplyr::filter(Y, lubridate::year(date) <= 2020 & lubridate::month(date) <= 2)

  }else if(freq == 'quarter'){

    Y.pre = dplyr::filter(Y, lubridate::year(date) < 2020)

  }

  # post-COVID shock
  if(freq == 'month'){

    Y.post = dplyr::filter(Y, lubridate::year(date) >= 2020 & lubridate::month(date) > 2)

    if(nrow(Y.post) < 4){stop('The COVID shock correction update requires four months starting 2020:M3.')}


  }else if(freq == 'quarter'){

    Y.post = dplyr::filter(Y, lubridate::year(date) >= 2020)

    if(nrow(Y.post) < 4){stop('The COVID shock correction update requires four quarters starting 2020:Q1.')}

  }

  # estimate volatility scalars -----------------
  theta =
    stats::optim(
      par = theta_initial,
      fn = cvc_likelihood,
      Y.pre = Y.pre,
      Y.post = Y.post,
      p = p,
      N = N
    )

  theta = as.vector(theta$par)

  # correct data -------------------------------

  # COIVD shock scale
  Y.pre$s = 1

  t = nrow(Y.post)
  j = c(0:(t-1))
  s_decay = rep(1, t) + (theta[3] - 1) * t(theta[4]^(j-2))
  s = c(theta[1:3], s_decay[4:t])

  Y.post$s = s

  Y = dplyr::bind_rows(Y.pre, Y.post)

  # correct the data
  Y.corrected = Y[, !colnames(Y) %in% c('s', 'date')] / Y[,'s']
  Y.corrected = data.frame(Y.corrected)
  Y.corrected$date = Y$date

  # re-estimate VAR ----------------------------

  var.corrected =
    sovereign::VAR(
      data = Y.corrected,
      horizon = var$model$horizon,
      freq = var$model$freq,
      type = var$model$type,
      p = var$model$p
    )

  var.corrected$correction.factors = theta

  # return corrected VAR -----------------------

  return(var.corrected)

}


# ##############################
#
# library(dplyr)             # general cleaning
# library(lubridate)         # date functions
#
# #-------------------------------------------
# # create data
# #-------------------------------------------
# # pull and prepare data from FRED
# quantmod::getSymbols.FRED(
#   c('UNRATE','INDPRO','GS10'),
#   env = globalenv())
#
# Data = cbind(UNRATE, INDPRO, GS10)
#
# Data = data.frame(Data, date = zoo::index(Data)) %>%
#   filter(lubridate::year(date) >= 1990) %>%
#   na.omit()
#
# var =
#   VAR(
#     data = Data,
#     horizon = 10,
#     freq = 'month',
#     lag.ic = 'BIC',
#     lag.max = 4)
#
# var = covid_volatility_correction(var)

